use_sim_time: false
roll:
  debugMode: false
  localizationMode: true
  mapUpdateEnabled: true

  # saveKeyframeMap: true
  # savePCD: true
  # savePose: true # for path recording purposes /reloc gt, save it to saveMapDirectory
  # saveLog: true
  # only test
  saveKeyframeMap: true
  savePCD: true
  savePose: true # for path recording purposes /reloc gt, save it to saveMapDirectory
  saveLog: true

  saveMapDirectory:  /home/lzq/ROLL_ws/logs/fastlio_noEdge/map_pcd        
  saveKeyframeMapDirectory: /home/lzq/ROLL_ws/logs/fastlio_noEdge/keyframes
  loadKeyframeMapDirectory: /home/lzq/ROLL_ws/logs/real/keyframes
#/home/lzq/ysc_ws/logs_nclt_fullmap/keyframes/keyframes
  saveRawCloud: false
  # # temporary mapping mode disabled
  # startTemporaryMappingInlierRatioThre: 0.0
  # temporary mapping mode 
  startTemporaryMappingInlierRatioThre: 0.3 #内点得分低于该值进入临时建图以及优化
  exitTemporaryMappingInlierRatioThre: 0.5 #内点得分高于该值退出，将LOAM优化得到结果应用到位姿
  slidingWindowSize: 30 #临时建图关键帧的滑动窗口数
  
  globalMatchingRate: 5

  alti0: 270.0
  lati: 42.293227
  longi0: -83.709657

  # initialGuess: [-0.023064531, 0.006521815, -0.123438711, 75.34987256, 108.1539385, -3.290657831] # 20120115
  # initialGuess: [-0.017984725855664, 0.012800135555456, -0.239421644654498,  75.520827533362393, 108.443935840162752, -3.291299069412723] # 20120202
  # initialGuess: [0.020559761872303, -0.000903047757892, -1.070181276478803, 75.579787639477303, 108.841026421505774, -3.316757230658669] # 20120317
  # initialGuess: [-0.015404291345797, -0.000618447067820, -0.312421191243955, 76.181700539534063, 108.837027977442645, -3.266038226333669] # 20120429
  # initialGuess: [0.019280939, 0.004926005, -1.755148225, -1.126196628, -3.182984251, 0.737359681] #20120511
  # initialGuess: [-0.054502938206381, -0.025535463484375, -0.192617339278269,88.661994612088748, 65.113330292655192, -2.807224100861569] #20120615
  # initialGuess: [-0.015221811894219, 0.001396211399287, -0.303854522173840,75.840364680427626, 109.017645300640666, -3.295762322731578] #20120804
  # initialGuess: [ -0.022510053101572, -0.009280409924894, -0.248550828547029, 77.430750494550978, 109.263558200969300, -3.198719816568997] #20121117
  # initialGuess: [-0.029788434570233, -0.010193528558576, -0.147575680110183,75.829339527800855, 107.724666286782551, -3.272894625813646] #20120110
  # initialGuess: [-0.008147915131152, -0.005539851055795, -0.519660742617843,78.049790086599444, 108.998587625038923, -3.169731476591604] #20130223

    # initialGuess: [ 0.004682509800383, 0.028480666632129, -0.634307081390563,76.505824066971385, 108.313730319190057, -3.255910820978710] # 20130405
  initialGuess: [0, 0, 0, 0, 0, 0] # 2024

  # Visualization
  globalMapVisualizationSearchRadius: 50.0    # meters, global map visualization radius(index would overflow if too big)
  globalMapVisualizationPoseDensity: 0.5       # 10 by default, meters, global map visualization pose density
  globalMapVisualizationLeafSize: 0.5          # 1 by default, meters, global map visualization cloud downsample size

    # Surrounding map
  surroundingkeyframeAddingDistThreshold: 1.0   # 1.0 by default,meters, regulate keyframe adding threshold 临时建图的关键帧判断距离
  surroundingkeyframeAddingAngleThreshold: 0.2  # 0.2 by default,radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 0.3               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 20         # 20 is not okay; meters, within n meters scan-to-map optimization (when loop closure disabled)
  # Topics
  pointCloudTopic: cloud_registered_body               # Point cloud data
  gpsTopic: fix1                  
  gtTopic: ground_truth

  # Frames
  lidarFrame: lidar_link
  mapFrame: odom

  # GPS Settings
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 1.0                      # m^2, threshold for using GPS data

  # NCLT HDL-32: 
  sensor: velodyne                            # lidar sensor type, either 'velodyne' or 'ouster'
  N_SCAN: 128                                 ## number of lidar channel (i.e., 16, 32, 64, 128)
  Horizon_SCAN: 1800                          ## lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048,robosense: 2000)
  downsampleRate: 1                           # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1

  lidarMinRange: 1                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 150.0                      ## default: 1000.0, maximum lidar range to be used

  # LOAM feature threshold
  edgeThreshold: 0.1
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometrySurfLeafSize: 0.4                     # default: 0.4 - outdoor, 0.2 - indoor
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

  # CPU Params
  numberOfCores: 4                              # number of cores for mapping optimization


  # Loop closure
  loopClosureEnableFlag: false
  loopClosureFrequency: 1.0                    # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # 30 by default,seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  # 0.3 by default, but for indoor it can go wrong, actually mostly for good matches, fitnessScore is around 0.03 ;
  #icp threshold, the smaller the better alignment
  historyKeyframeFitnessScore: 0.2              

